index/scorch/scorch.go

```diff
@@ -266,15 +267,18 @@ func (s *Scorch) Batch(batch *index.Batch) error {
 
 	atomic.AddUint64(&s.stats.analysisTime, uint64(time.Since(start)))
 
-	// notify handlers that we're about to introduce a segment
-	s.fireEvent(EventKindBatchIntroductionStart, 0)
-
 	var newSegment segment.Segment
 	if len(analysisResults) > 0 {
-		newSegment = mem.NewFromAnalyzedDocs(analysisResults)
+		newSegment, err = zap.NewSegmentBase(mem.NewFromAnalyzedDocs(analysisResults), DefaultChunkFactor)
+		if err != nil {
+			return err
+		}
 	}
 
-	err := s.prepareSegment(newSegment, ids, batch.InternalOps)
+	// notify handlers that we're about to introduce a segment
+	s.fireEvent(EventKindBatchIntroductionStart, 0)
```

Comment:
wondering why you moved the firing of this event, i believe this event is used to block creating new segments by the herder, and with this change, we will create segments before firing the event, and so we've already created the memory hungry structure before blocking...

Reply:
Ah, I just read that EventKindBatchIntroductionStart and thought that the event was fired too early -- we're not actually trying to hand the segment to the introducer yet at that point?

Also, more on the memory thoughts... at this point, analysis is already done so we're already holding onto a bunch of memory for the analysis results.  By going all the way to the zap-based representation, it should hopefully (not measured!) be a smaller representation that we hold onto when EventKindBatchIntroductionStart blocks?

Given that, though, I can swap the lines back in a bit.

Reply:
Ya but we created the larger mem representation already and turned into garbage.  So, it seems like, for a scenario in which we wanted to wait and not make the system worse, we actually did some damage.  It means there could be as many as 1024 segments already build, but not introduced.  What is the point in blocking them now, they take up memory but are not searchable.  It seems to me we should block before doing anything, so perhaps that is before analysis...

Reply:
Makes sense -- I think the best tactical step right now is I'll swap the lines back.

And the thought of perhaps moving the blocking to before analysis should be a separate followup thread for another day / different PR.